zulip_bots/zulip_bots/bots/gameoffifteen/gameoffifteen.py

```diff
@@ -0,0 +1,148 @@
+import copy
+import random
+
+from typing import List, Any, Tuple
+from zulip_bots.game_handler import GameAdapter, BadMoveException
+
+class GameOfFifteenModel(object):
+
+    final_board = [[0, 1, 2],
+                   [3, 4, 5],
+                   [6, 7, 8]]
+
+    initial_board = [[8, 7, 6],
+                     [5, 4, 3],
+                     [2, 1, 0]]
+
+    def __init__(self, board: Any=None) -> None:
+        if board is not None:
+            self.current_board = board
+        else:
+            self.current_board = copy.deepcopy(self.initial_board)
+
+    def determine_game_over(self, players: List[str]) -> str:
+        if self.won(self.current_board):
+            return 'current turn'
+        return ''
+
+    def won(self, board: Any) -> bool:
+        for i in range(3):
+            for j in range(3):
+                if (board[i][j] != self.final_board[i][j]):
+                    return False
+        return True
+
+    def validate_move(self, tile: int) -> bool:
+        if tile < 1 or tile > 8:
+            return False
+        return True
+
+    def update_board(self, board):
+        self.current_board = copy.deepcopy(board)
+
+    def make_move(self, move: str, player_number: int, computer_move: bool=False) -> Any:
+        board = self.current_board
+        move = move.strip()
+        move = move.split(' ')
+        if '' in move:
+            raise BadMoveException('You should enter space separated digits.')
+        moves  = len(move)
+        flag = 0
+        for m in range(1, moves):
+            tile = int(move[m])
+            flag = 0
+            for i in range(3):
+                for j in range(3):
+                    if board[i][j] == tile:
```

Comment:
We should have a data structure that maps tiles to their coordinates, so that we don't need to do this nested loop just to find a tile.